File: leetcode/337.打家劫舍 III/337.打家劫舍 III.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def rob(self, root: TreeNode) -> int:
        memo = {}

        def do_rob(root: TreeNode) -> int:
            if root is None:
                return 0
            if root in memo:
                return memo[root]

            rob_val_l, rob_val_r = 0, 0
            if root.left is not None:
                # 因为是二叉树，下下个节点 分别为 根左左，根左右
                rob_val_l = do_rob(root.left.left) + do_rob(root.left.right)
            if root.right is not None:
                # 因为是二叉树，下下个节点 分别为 根右左，根右右
                rob_val_r = do_rob(root.right.left) + do_rob(root.right.right)
            # 打劫此节点 + 下下个节点
            do_it = root.val + rob_val_l + rob_val_r
            # 不打劫此节点+ 打劫下个节点
            not_do_it = do_rob(root.left) + do_rob(root.right)
            # 取出最有价值的路线
            res = max(do_it, not_do_it)
            # 存入dp数组
            memo[root] = res
            return res

        return do_rob(root)

File: leetcode/337.打家劫舍 III/337.打家劫舍 III.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func rob(root *TreeNode) int {
	memo := make(map[*TreeNode]int)
	return doRob(root, memo)
}

func doRob(root *TreeNode, memo map[*TreeNode]int) int {
	if root == nil {
		return 0
	}
	if val, ok := memo[root]; ok {
		return val
	}
	doItL := 0
	doItR := 0
	if root.Left != nil{
		doItL = doRob(root.Left.Left, memo) + doRob(root.Left.Right, memo)
	}
	if root.Right != nil{
		doItR = doRob(root.Right.Left, memo) + doRob(root.Right.Right, memo)
	}
	doIt := root.Val + doItL + doItR
	notDoIt := doRob(root.Left, memo) + doRob(root.Right, memo)
	res := max(doIt, notDoIt)
	memo[root] = res
	return res
}

func max(int1 int, int2 int) int {
	if int1 < int2 {
		return int2
	} else {
		return int1
	}
}
